[ARM64] Use a different crash code for brk on ASAN builds.

MenloDorian opened the pull request:
#### 370e7cba9f6ab189d85746ea028d46528a57a74e
<pre>
[ARM64] Use a different crash code for brk on ASAN builds.
<a href="https://bugs.webkit.org/show_bug.cgi?id=271795">https://bugs.webkit.org/show_bug.cgi?id=271795</a>
<a href="https://rdar.apple.com/125494279">rdar://125494279</a>

Reviewed by Yusuke Suzuki and Justin Michaud.

* Source/JavaScriptCore/assembler/MacroAssemblerARM64.cpp:
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::breakpoint):
* Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h:
(JSC::MacroAssemblerRISCV64::breakpoint):
* Source/JavaScriptCore/llint/LowLevelInterpreter.cpp:
* Source/JavaScriptCore/offlineasm/arm64.rb:
* Source/WTF/wtf/Assertions.cpp:
* Source/WTF/wtf/Assertions.h:
* Source/bmalloc/libpas/src/libpas/pas_utils.c:

Canonical link: <a href="https://commits.webkit.org/276776@main">https://commits.webkit.org/276776@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c58accc838ae1340a66586ccc5fc4576e532390

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45618 "15 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24744 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48286 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41653 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22140 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37366 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46196 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21817 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39352 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18507 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19218 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40441 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3662 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/38846 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40771 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50028 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/45086 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20606 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17127 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44457 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21912 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43293 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22272 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/52243 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6354 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21599 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10677 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
r=me

kmiller68 commented on Source/WTF/wtf/Assertions.h:
```diff
@@ -268,9 +268,15 @@ WTF_EXPORT_PRIVATE bool WTFIsDebuggerAttached(void);
+#else
+#define CRASH_CODE 0xc471
 #define CRASH_INST "brk #0xc471"
+#endif
```
It seems like these could be guarded on the value not being defined already?

MenloDorian replied:
I think if there's a collision, we want it to error out.  So, I think keeping it as is would work.  Duplicate definitions will result in an error.

kmiller68 replied:
I think the idea is that some other project wants to set their own `CRASH_CODE`. Right now they can't without modifying the source code. if this was
```c++
#if !defined(CRASH_CODE)
#if ASAN_ENABLED
#define CRASH_CODE 0x0
#else
#define CRASH_CODE 0xc471
#endif
#endif

#if !defined(CRASH_INST)
#define CRASH_INST "brk #" STRINGIZE_VALUE_OF(CRASH_CODE)
#endif
``` 
They would be able to.

kmiller68 replied:
Alternatively, the folks we're making this change for could set those values so we could just do:
```c++
#if !defined(CRASH_CODE)
#define CRASH_CODE 0xc471
#endif

#if !defined(CRASH_INST)
#define CRASH_INST "brk #" STRINGIZE_VALUE_OF(CRASH_CODE)
#endif
```

MenloDorian replied:
It is important that we use 0xc471 in production.  Hence, I would not want an option to be able to silently replace it with something else.  So, if it's needed, they should change the code ... especially since that is rare.

kmiller68 replied:
Almost any feature flag can be overridden by setting a value before building causing who knows what problems. I don't think this is much different.

MenloDorian replied:
Talked with Keith about this offline.  His main concern is to empower custom builds like ASAN to specify whatever crash code they want in case they want some specific one.  I think it is rare that this would be needed, and I fear some innocent code accidentally overriding it silently.  

However, I can accommodate this with reduced risk by renaming CRASH_CODE to something like WTF_FATAL_CRASH_CODE, which is less likely to be defined willy nilly by random code.

justinmichaud replied:
I think we should err on the side of making ASAN folks change source code rather than risk a random security regression.

kmiller68 replied:
If we’re going to assume someone can set these values why couldn’t  someone just set ASAN_ENABLED to 1?

Constellation reviewed (approved):
r=me too

Commit:
[ARM64] Use a different crash code for brk on ASAN builds.
https://bugs.webkit.org/show_bug.cgi?id=271795
rdar://125494279

Reviewed by Yusuke Suzuki and Justin Michaud.

* Source/JavaScriptCore/assembler/MacroAssemblerARM64.cpp:
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::breakpoint):
* Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h:
(JSC::MacroAssemblerRISCV64::breakpoint):
* Source/JavaScriptCore/llint/LowLevelInterpreter.cpp:
* Source/JavaScriptCore/offlineasm/arm64.rb:
* Source/WTF/wtf/Assertions.cpp:
* Source/WTF/wtf/Assertions.h:
* Source/bmalloc/libpas/src/libpas/pas_utils.c:

Canonical link: https://commits.webkit.org/276776@main
